server/handlers/authentication.go

```diff
@@ -50,6 +50,9 @@ func (a *authenticationHandlerImpl) getLoginHandler(w http.ResponseWriter, r *ht
 			http.Redirect(w, r, next, http.StatusSeeOther)
 			return
 		}
+	} else if token == nil && err != nil {
```

Comment:
```
token, err := r.Cookie("token")
if err != nil {
     a.parseLoginWithErr(w, r, err)
     return
}

user, err := a.service.GetUserByToken(token.Value)
if err != nil {
      a.parseLoginWithErr(w, r, err)
      return
}

http.Redirect(w, r, next, http.StatusSeeOther)

Reply:
This way doesn't work, it gives the following error in unit test:

```error
--- FAIL: Test_GetLoginHandler_TokenExists_OK (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x783e6f]

goroutine 114 [running]:
testing.tRunner.func1.2({0x7e60a0, 0xb9ddc0})
	/usr/local/go/src/testing/testing.go:1631 +0x24a
testing.tRunner.func1()
	/usr/local/go/src/testing/testing.go:1634 +0x377
panic({0x7e60a0?, 0xb9ddc0?})
	/usr/local/go/src/runtime/panic.go:770 +0x132
globe-and-citizen/layer8/server/handlers.(*authenticationHandlerImpl).getLoginHandler(0xc00042cf48, {0x91aec8, 0xc0001916c0}, 0xc000306480)
	/home/hmk/Hmk/GlobeCitizen/l8-local-test/layer8/server/handlers/authentication.go:59 +0x28f
globe-and-citizen/layer8/server/handlers.(*authenticationHandlerImpl).Login(0xc000306480?, {0x91aec8?, 0xc0001916c0?}, 0x6?)
	/home/hmk/Hmk/GlobeCitizen/l8-local-test/layer8/server/handlers/authentication.go:32 +0x6e
globe-and-citizen/layer8/server/handlers_test.Test_GetLoginHandler_TokenExists_OK(0xc0001da340)
	/home/hmk/Hmk/GlobeCitizen/l8-local-test/layer8/server/handlers/authentication_test.go:70 +0x326
testing.tRunner(0xc0001da340, 0x8a47b8)
	/usr/local/go/src/testing/testing.go:1689 +0xfb
created by testing.(*T).Run in goroutine 1
	/usr/local/go/src/testing/testing.go:1742 +0x390
FAIL	globe-and-citizen/layer8/server/handlers	0.015s
```

Reply:
Interesting.  How about this: 
```
func (a *authenticationHandlerImpl) getLoginHandler(w http.ResponseWriter, r *http.Request) {
	next := r.URL.Query().Get("next")
	if next == "" {
		next = "/"
	}
	// check if the user is already logged in
	token, err := r.Cookie("token")
	if err != nil {
		a.parseHTML(w, "assets-v1/templates/src/pages/oauth_portal/login.html", map[string]interface{}{
			"HasNext":  next != "",
			"Next":     next,
			"ProxyURL": os.Getenv("PROXY_URL"),
		},
		)
		return
	}

	user, err := a.service.GetUserByToken(token.Value)
	if err != nil {
		a.parseLoginWithErr(w, r, err)
		return
	}

	if user == nil {
		a.parseHTML(w, "assets-v1/templates/src/pages/oauth_portal/login.html", map[string]interface{}{
			"HasNext":  next != "",
			"Next":     next,
			"ProxyURL": os.Getenv("PROXY_URL"),
		},
		)
	}

	http.Redirect(w, r, next, http.StatusSeeOther)

	return
}
```

Reply:
```Go
=== RUN   Test_GetLoginHandler_NoToken_OK
    authentication_test.go:36: 
        	Error Trace:	/home/hmk/Hmk/GlobeCitizen/l8-local-test/layer8/server/handlers/authentication_test.go:36
        	            				/home/hmk/Hmk/GlobeCitizen/l8-local-test/layer8/server/handlers/authentication.go:75
        	            				/home/hmk/Hmk/GlobeCitizen/l8-local-test/layer8/server/handlers/authentication.go:32
        	            				/home/hmk/Hmk/GlobeCitizen/l8-local-test/layer8/server/handlers/authentication_test.go:46
        	Error:      	Not equal: 
        	            	expected: map[string]interface {}{"Error":"http: named cookie not present", "HasNext":true, "Next":""}
        	            	actual  : map[string]interface {}{"HasNext":true, "Next":"/", "ProxyURL":"http://localhost:5001"}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,5 +1,5 @@
        	            	 (map[string]interface {}) (len=3) {
        	            	- (string) (len=5) "Error": (string) (len=30) "http: named cookie not present",
        	            	  (string) (len=7) "HasNext": (bool) true,
        	            	- (string) (len=4) "Next": (string) ""
        	            	+ (string) (len=4) "Next": (string) (len=1) "/",
        	            	+ (string) (len=8) "ProxyURL": (string) (len=21) "http://localhost:5001"
        	            	 }
        	Test:       	Test_GetLoginHandler_NoToken_OK
    authentication_test.go:49: 
        	Error Trace:	/home/hmk/Hmk/GlobeCitizen/l8-local-test/layer8/server/handlers/authentication_test.go:49
        	Error:      	Not equal: 
        	            	expected: 401
        	            	actual  : 200
        	Test:       	Test_GetLoginHandler_NoToken_OK
```

I think the one we currently have is working fine with unit tests, I testes these two (your's and Firhan's) but both are breaking the unit tests.

Reply:
Sorry for wrong suggestion, this one wont break the test :

```
func (a *authenticationHandlerImpl) getLoginHandler(w http.ResponseWriter, r *http.Request) {
	next := r.URL.Query().Get("next")
	if next == "" {
		next = "/"
	}

	// check if the user is already logged in
	token, err := r.Cookie("token")
	if err != nil && token == nil {
		a.parseLoginWithErr(w, r, err)
		return
	}

	user, err := a.service.GetUserByToken(token.Value)
	if err != nil {
		a.parseLoginWithErr(w, r, err)
		return
	}

	if user != nil {
		http.Redirect(w, r, next, http.StatusSeeOther)
		return
	}

	a.parseHTML(w, http.StatusOK, "assets-v1/templates/src/pages/oauth_portal/login.html",
		map[string]interface{}{
			"HasNext":  next != "",
			"Next":     next,
			"ProxyURL": os.Getenv("PROXY_URL"),
		},
	)
}
```